pygmt/src/grdhisteq.py

```diff
@@ -0,0 +1,358 @@
+        list of data values that divide the data range into divisions which
+        have an equal area in the image [Default is 16 if ``divisions`` is not
+        set]. The :class:`pandas.DataFrame` or ASCII file output can be used to
+        make a colormap with :meth:`pygmt.makecpt` and an image with
+        :meth:`pygmt.Figure.grdimage` that has all levels of gray occuring
+        equally.
+
+        :meth:`pygmt.grdhisteq` also provides a way to write a grid with
+        statistics based on a cumulative distribution function. In this
+        application, the ``outgrid`` has relative highs and lows in the same
+        (x,y) locations as the ``grid``, but the values are changed to reflect
+        their place in the cumulative distribution.
+
+        Must provide ``outfile`` or ``outgrid``.
+
+        Full option list at :gmt-docs:`grdhisteq.html`
+
+        {aliases}
+
+        Parameters
+        ----------
+        grid : str or xarray.DataArray
+            The file name of the input grid or the grid loaded as a DataArray.
+        outgrid : str or bool or None
+            The name of the output netCDF file with extension .nc to store the
+            grid in.
+        outfile : str or bool or None
+            The name of the output ASCII file to store the results of the
+            histogram equalization in. Not allowed if ``outgrid`` is used.
+        divisions : int
+            Set the number of divisions of the data range [Default is 16].
+
+        {R}
+        {V}
+
+        Returns
+        -------
+        ret: pandas.DataFrame or xarray.DataArray or None
+            Return type depends on whether the ``outgrid`` parameter is set:
+
+            - pandas.DataFrame if ``outfile`` is True
+            - xarray.DataArray if ``outgrid`` is True
+            - None if ``outgrid`` is a str (grid output is stored in
+              ``outgrid``)
+            - None if ``outfile`` is a str (file output is stored in
+              ``outfile``)
+
+        See Also
+        -------
+        :meth:`pygmt.grd2cpt`
+        """
+
+        with Session() as lib:
+            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
+            with file_context as infile:
+                arg_str = " ".join([infile, build_arg_string(kwargs)])
+                lib.call_module("grdhisteq", arg_str)
+
+        if output_type == "file":
+            return None
+        if output_type == "xarray":
+            return load_dataarray(tmpfile.name)
+
+        result = pd.read_csv(
+            filepath_or_buffer=kwargs["D"],
+            sep="\t",
+            header=None,
+            names=["start", "stop", "bin_id"],
+            dtype={
+                "start": np.float32,
+                "stop": np.float32,
+                "bin_id": np.uint32,
+            },
+        )
+        if output_type == "numpy":
+            result = result.to_numpy()
```

Comment:
Need to add a test for `compute_bins(..., output_type="numpy")`. Can probably turn `test_compute_bins_no_outfile` into a parametrized test (or just make a new one).